crypto/x509/x509_cmp.c

```diff
@@ -251,20 +251,21 @@ int X509_NAME_cmp(const X509_NAME *a, const X509_NAME *b)
         return -1;
 
     /* Ensure canonical encoding is present and up to date */
-    if (!a->canon_enc || a->modified) {
+    if (a->canon_enc == NULL || a->modified) {
         ret = i2d_X509_NAME((X509_NAME *)a, NULL);
         if (ret < 0)
             return -2;
     }
 
-    if (!b->canon_enc || b->modified) {
+    if (b->canon_enc == NULL || b->modified) {
         ret = i2d_X509_NAME((X509_NAME *)b, NULL);
         if (ret < 0)
             return -2;
     }
 
     ret = a->canon_enclen - b->canon_enclen;
-    if (ret == 0 && a->canon_enclen != 0)
+    if (ret == 0 && a->canon_enclen != 0
+            && a->canon_enc != NULL && b->canon_enc != NULL)
         ret = memcmp(a->canon_enc, b->canon_enc, a->canon_enclen);
```

Comment:
So this change means we skip the memcmp and return 0 if `canon_enclen >0` but `canon_enc == NULL`. Surely though that scenario is an error (and so a -2 return is more appropriate)?

Reply:
That's more sensible.  I've made the change.

Reply:
@mattcaswell and @pauli, this 'more sensible' solution turns out to cause trouble.
I found it is the reason for #14813.
The point is that NULL-DNs have zero `canon_enclen` and `canon_enc` is NULL, 
and the latter must not be taken as an error in case  `canon_enclen` is 0.
So the original plain to fix the Coverity issue was actually better.

It suppose the Coverity issue is a false positive 
because  `canon_enclen != 0` should imply `canon_enc != NULL`.

Reply:
IMO the actual issue/but is that for a NULL-DN the `canon_enc` should be a (static) empty string, which should also avoid artificial case distinctions. I'll fix it this way.

Reply:
You're more expert than I, PR welcome :)
If not, I'll look into it next week (assuming I remember, which is doubtful :)

Reply:
Well, my original plan was to leave this to an ASN.1 expert,
but this can be fixed without touching the actual ASN.1 stuff :)